### PR TITLE
Code optimizations reported by static code analysis (2020-02-18)

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

SUMMARY: Performance "Code optimizations reported by static code analysis"

#### Purpose of change

Cleanup code.

#### Describe the solution

See commit messages.

### CODE REVIEW

**Comment by pjf** [comment]
As a matter of personal interest and education, what are we using for static code analysis? Is it `clang -analyze`, or are these better/different options these days?

**Comment by ZhilkinSerg** [comment]
PVS Studio, ReSharper C++

**Comment by pjf** [comment]
Oh, I've heard good things about ReSharper, but didn't realise it did C++ as well as C#. Thank you, @ZhilkinSerg !

File: src/activity_item_handling.cpp [hunk 1]
```diff
@@ -1992,12 +1991,12 @@ static void fetch_activity( player &p, const tripoint &src_loc,
...
                     it.set_var( "activity_var", p.name );
                     p.i_add( it );
-                    picked_up = it.tname();
                     items_there.erase( item_iter );
                     // If we didn't pick up a whole stack, put the remainder back where it came from.
                     if( leftovers.charges > 0 ) {
                         g->m.add_item_or_charges( src_loc, leftovers );
                     }
+                    std::string picked_up = it.tname();
                     if( p.is_npc() && !picked_up.empty() ) {
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: Under what circumsntance would `item::tname` return an empty string? It seems this check will never be false.

Comment (ZhilkinSerg) [inline src/activity_item_handling.cpp]:
```suggestion
                    if( p.is_npc() ) {
```

File: src/activity_item_handling.cpp [hunk 2]
```diff
@@ -1992,12 +1991,12 @@ static void fetch_activity( player &p, const tripoint &src_loc,
                     }
                     it.set_var( "activity_var", p.name );
                     p.i_add( it );
-                    picked_up = it.tname();
                     items_there.erase( item_iter );
                     // If we didn't pick up a whole stack, put the remainder back where it came from.
                     if( leftovers.charges > 0 ) {
                         g->m.add_item_or_charges( src_loc, leftovers );
                     }
+                    std::string picked_up = it.tname();
```
Comment (ZhilkinSerg) [inline src/activity_item_handling.cpp]:
```suggestion
```

File: src/activity_item_handling.cpp [hunk 3]
```diff
@@ -1992,12 +1991,12 @@ static void fetch_activity( player &p, const tripoint &src_loc,
...
                     if( p.is_npc() && !picked_up.empty() ) {
                         if( pickup_count == 1 ) {
                             add_msg( _( "%1$s picks up a %2$s." ), p.disp_name(), picked_up );
```
Comment (ZhilkinSerg) [inline src/activity_item_handling.cpp]:
```suggestion
                            add_msg( _( "%1$s picks up a %2$s." ), p.disp_name(), it.tname() );
```

**Pull request merged**